SingularityService/src/main/java/com/hubspot/singularity/data/history/SingularityTaskHistoryPersister.java

```diff
@@ -65,43 +56,36 @@ public void runActionOnPoll() {
       LOG.info("Checking inactive task ids for task history persistence");
 
       final long start = System.currentTimeMillis();
-      final List<SingularityTaskId> allTaskIds = taskManager.getAllTaskIds();
-      allTaskIds.sort(SingularityTaskId.STARTED_AT_COMPARATOR_DESC);
-
-      AtomicInteger numTotal = new AtomicInteger();
-      AtomicInteger numTransferred = new AtomicInteger();
-
-      for (String requestId : requestManager.getAllRequestIds()) {
-        lock.runWithRequestLock(() -> {
-          List<SingularityTaskId> activeForRequest = taskManager.getActiveTaskIdsForRequest(requestId);
-          Set<SingularityTaskId> lbCleaningTaskIds = Sets.newHashSet(taskManager.getLBCleanupTasks());
-          List<SingularityPendingDeploy> pendingDeploys = deployManager.getPendingDeploys();
-
-          long startRequest = System.currentTimeMillis();
-          AtomicInteger transferred = new AtomicInteger();
-          for (SingularityTaskId taskId : allTaskIds) {
-            if (taskId.getRequestId().equals(requestId) &&
-                !(activeForRequest.contains(taskId) || lbCleaningTaskIds.contains(taskId) || isPartOfPendingDeploy(pendingDeploys, taskId) || couldReturnWithRecoveredAgent(taskId))
-            ) {
-              if (moveToHistoryOrCheckForPurge(taskId, transferred.getAndIncrement())) {
-                numTransferred.getAndIncrement();
-              }
-
-              numTotal.getAndIncrement();
-              if (System.currentTimeMillis() - startRequest > maxPersisterLockTimeMillis) {
-                break;
-              }
-            }
+      Map<String, List<SingularityTaskId>> inactiveTaskIdsByRequest = getInactiveTaskIdsByRequest();
+      for (Map.Entry<String, List<SingularityTaskId>> entry : inactiveTaskIdsByRequest.entrySet()) {
+        int forRequest = 0;
+        int transferred = 0;
+        for (SingularityTaskId taskId : entry.getValue()) {
+          if (moveToHistoryOrCheckForPurge(taskId, forRequest)) {
+            transferred++;
           }
-        }, requestId, "task history persister");
+
+          forRequest++;
+        }
+        LOG.info("Transferred {} out of {} inactive task ids in {}", transferred, entry.getValue().size(), JavaUtils.duration(start));
       }
 
-      LOG.info("Transferred {} out of {} inactive task ids (total {}) in {}", numTransferred, numTotal, allTaskIds.size(), JavaUtils.duration(start));
     } finally {
       persisterLock.unlock();
     }
   }
 
+  private Map<String, List<SingularityTaskId>> getInactiveTaskIdsByRequest() {
+    final List<SingularityTaskId> taskIds = new ArrayList<>(taskManager.getAllTaskIds());
+    taskIds.removeAll(taskManager.getActiveTaskIds());
+    taskIds.removeAll(taskManager.getLBCleanupTasks());
```

Comment:
Not a huge deal, but can we do these `removeAll`s against a Set instead of a List?

Reply:
yeah definitely, originally had it as a list because i was sorting it, but then switched to the stream at the end instead. Will update